server/services/v1/auth.go

```diff
@@ -42,53 +52,194 @@ type authService struct {
+	// validate
+	client, err := a.management.Client.Read(req.GetId())
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to delete application: reason = %s", err.Error())
+	}
+
+	// check ownership before deleting
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to delete application: reason = %s", err.Error())
+	}
+	if client.GetClientMetadata()[createdBy] != currentSub {
+		return nil, api.Errorf(api.Code_PERMISSION_DENIED, "Failed to delete application: reason = You cannot delete application that is not created by you.")
+	}
+
+	err = deleteApplication(a, req.GetId())
+	if err != nil {
+		log.Warn().Msgf("Failed to cleanup half-created app with clientId=%s, reason=%s", req.GetId(), err.Error())
+		return nil, err
+	}
+	return &api.DeleteApplicationResponse{
+		Deleted: true,
+	}, nil
+}
+
+func (a *Auth0) RotateApplicationSecret(ctx context.Context, req *api.RotateApplicationSecretRequest) (*api.RotateApplicationSecretResponse, error) {
+	// validate
+	client, err := a.management.Client.Read(req.GetId())
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to rotate application secret: reason = %s", err.Error())
+	}
+
+	// check ownership before rotating
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to delete application: reason = %s", err.Error())
+	}
+	if client.GetClientMetadata()[createdBy] != currentSub {
+		return nil, api.Errorf(api.Code_PERMISSION_DENIED, "Failed to rotate application secret: reason = You cannot rotate secret for application that is not created by you.")
+	}
+
+	updatedApp, err := a.management.Client.RotateSecret(req.GetId())
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to rotate application secret: reason = %s", err.Error())
+	}
+	return &api.RotateApplicationSecretResponse{
+		Application: &api.Application{
+			Id:          updatedApp.GetClientID(),
+			Name:        updatedApp.GetName(),
+			Description: updatedApp.GetDescription(),
+			Secret:      updatedApp.GetClientSecret(),
+			CreatedAt:   readCreatedAt(updatedApp),
+			CreatedBy:   updatedApp.GetClientMetadata()[createdBy],
+		},
+	}, nil
+}
+
+func (a Auth0) ListApplications(ctx context.Context, _ *api.ListApplicationsRequest) (*api.ListApplicationsResponse, error) {
+
+	appList, err := a.management.Client.List(management.IncludeFields("client_id", "client_metadata", "client_secret", "description", "name"))
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
+	}
+
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
+	}
+
+	var apps []*api.Application
+	for _, client := range appList.Clients {
+		// filter for this user's apps
```

Comment:
Can a user have more than one app?

Reply:
Also, if it is for a single client then is it possible to issue a `list` on that client?

Reply:
I have reached out to auth0 on how to filter in this query for the metadata attribute. I am waiting for them.

Reply:
sg